enhance(main/cmus): Add AAudio output plugin

pgaskin opened the pull request:
Support native audio on Android 8.1 and later.

- Better battery life.
- Configurable latency.
- Configurable capture policy.
- Configurable sharing mode.
- Configurable spatialization.
- Full multi-channel audio support.
- Supports `pause_on_output_change`.

https://github.com/cmus/cmus/pull/1346
https://github.com/cmus/cmus/commit/3880d9df1c03e8931a59a3dd22daade8457d1b43

TomJo2000 commented:
We'd likely prefer to wait for the above mentioned PR in `cmus` to:
1. Be merged, and
2. Be shipped as part of a release.

See "[Don't ship WIP](https://dont-ship.it/)" for a more detailed breakdown.
If we do want to ship it before a new release that would require additional coordination with upstream.

TomJo2000 commented:
Ah okay thank you that clears things up.

TomJo2000 commented:
Do you mind if I do a review pass while the PR is still in the draft stage?
I'm thinking we can avoid needing to inline the patch into the termux-packages repository by adding it as a second item to `TERMUX_PKG_SRCURL` (as well as a corresponding `TERMUX_PKG_SHA256`) in the build script.

pgaskin commented:
Yeah, feel free to do a review. I just got the package build to work, and I've already been testing the plugin for a few days, so the code's pretty much final.

pgaskin commented on packages/cmus/aaudio.patch:
```diff
@@ -0,0 +1,994 @@
++	{ NULL },
++};
++
++const int op_priority = -3; // higher priority than pulse (-2)
```
I set this to be a higher priority than pulse since it should almost always be the better option if it's supported.

Another option could be to split the pulseaudio plugin into a subpackage, with the aaudio plugin having a lower priority.

TomJo2000 reviewed (commented):
We should be able to do something along the lines of what the `bash` package does.
https://github.com/termux/termux-packages/blob/58ee4ec7004061ad8f679853a8bc06cc67382dc6/packages/bash/build.sh#L39-L78

Although, since this is a single patch we're not going to need the whole array shenanigans.

pgaskin commented on packages/cmus/aaudio.patch:
```diff
@@ -0,0 +1,994 @@
++}
++"
++
++check_aaudio()
++{
++	if check_library AAUDIO "" "-laaudio"
++	then
++		AAUDIO_CFLAGS="$AAUDIO_CFLAGS -D__ANDROID_UNAVAILABLE_SYMBOLS_ARE_WEAK__ -Werror=unguarded-availability"
++
++		msg_checking "for aaudio support"
++		if try_compile_link "$aaudio_code" $AAUDIO_CFLAGS $AAUDIO_LIBS
++		then
++			msg_result yes
++			return 0
++		fi
++		msg_result no
++	fi
++	return 1
++}
```
I will need to change this logic since it isn't adding the linker flags when compiled with the package build.sh, probably due to the weak linking causing it to succeed without the `-laaudio`.

TomJo2000 replied:
We could always add `LDFLAGS+=' -laaudio'` in the preconfigure step.

pgaskin replied:
Only the output plugin should be linked to audio, otherwise, cmus won't work on versions without it anymore.

TomJo2000 replied:
Ah yep that makes sense.

pgaskin replied:
I think I've fixed the issue; it was due to weak linking and the toolchain config.

pgaskin commented:
- Changed the build script to download the patch from GitHub.
- Fixed the link error due to Termux using `-Wl,--as-needed` and libaaudio using weak symbols.
- Did some minor refactoring of the AAudio output plugin.

The build failure is unrelated to this change (it was fine when I tested it just now).

pgaskin commented:
The build failure is caused by `clang: error: unknown argument: '-fno-openmp-implicit-rpath'` because `pkg-config --libs ncurses` is `-L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw` (it was `-lncurses -ltinfo` 4 days ago).

Biswa96 commented:
> The build failure is caused by `clang: error: unknown argument: '-fno-openmp-implicit-rpath'`

Please wait for #21038

TomJo2000 commented on packages/cmus/build.sh:
`TERMUX_PKG_REVISION=1` should be set below the version.

TomJo2000 reviewed (approved):
Couple nits, but no showstoppers.
Thanks a lot of putting in the effort.

TomJo2000 reviewed (approved):
Lgtm

Biswa96 commented:
No idea 😔

TomJo2000 commented:
@pgaskin thank you for your contribution.
